pkg/agent/flowexporter/exporter/l7Listener.go

```diff
@@ -0,0 +1,184 @@
+		return
+	}
+	defer listener.Close()
+
+	klog.InfoS("L7 Listener Server started. Listening for connections...")
+	for {
+		select {
+		case <-stopCh:
+			return
+		default:
+			conn, err := listener.Accept()
+			if err != nil {
+				klog.ErrorS(err, "Error accepting Suricata connection")
+				continue
+			}
+			go l.handleClientConnection(conn)
+		}
+	}
+}
+
+func (l *l7Listener) handleClientConnection(conn net.Conn) {
+	defer conn.Close()
+	reader := bufio.NewReader(conn)
+	for {
+		buffer, err := reader.ReadBytes('\n')
+		if err != nil && err != io.EOF {
+			klog.ErrorS(err, "Error reading data", "buffer", buffer)
+			return
+		}
+		err = l.processLog(buffer)
+		if err != nil {
+			klog.ErrorS(err, "Error while processing L7 data")
+			return
+		}
+	}
+}
+
+func (l *l7Listener) processLog(data []byte) error {
+	l.l7mut.Lock()
+	defer l.l7mut.Unlock()
+	var eventType string
+	// Check if the event type is "http"
+	if eventType = extractEventType(data); eventType != "http" {
+		return nil
+	}
+	var event JsonToEvent
+	err := json.Unmarshal(data, &event)
+	if err != nil {
+		return fmt.Errorf("Error parsing JSON data %v", data)
+	}
+	if err = l.addOrUpdateL7EventMap(&event, eventType); err != nil {
+		return fmt.Errorf("Error while adding or Updating L7 event map %v", err)
+	}
+	return nil
+}
+
+func extractEventType(data []byte) string {
+	// Define a temporary struct for extracting the event type
+	type Temp struct {
+		EventType string `json:"event_type"`
+	}
+
+	// Parse the JSON string and extract the event type
+	var temp Temp
+	err := json.Unmarshal(data, &temp)
+	if err != nil && !strings.Contains(err.Error(), "unexpected end of JSON input") {
+		klog.ErrorS(err, "Error parsing JSON for event type:", "data", data)
+		return ""
+	}
+
+	return temp.EventType
+}
+
+func (l *l7Listener) addOrUpdateL7EventMap(event *JsonToEvent, eventType string) error {
+	protocol, err := flowexporter.LookupProtocolMap(event.Proto)
+	if err != nil {
+		return fmt.Errorf("Invalid Protocol type, err: %v", err)
+	}
+	conn := flowexporter.Connection{
+		FlowKey: flowexporter.Tuple{
+			SourceAddress:      event.SrcIP,
+			DestinationAddress: event.DestIP,
+			Protocol:           protocol,
+			SourcePort:         uint16(event.SrcPort),
+			DestinationPort:    uint16(event.DestPort),
+		},
+	}
+	connKey := flowexporter.NewConnectionKey(&conn)
+
+	switch eventType {
+	case "http":
+		if _, ok := l.l7EventsHttp[connKey]; !ok {
+			l.l7EventsHttp[connKey] = make(map[int32]Http)
+		}
+		l.l7EventsHttp[connKey][event.TxID] = event.HTTP
+	}
+	return nil
+}
+
+func (l *l7Listener) getL7EventDataHttpVals(connkey flowexporter.ConnectionKey, destinationPodNN, sourcePodNN string, exp *FlowExporter) string {
+	l.l7mut.Lock()
+	defer l.l7mut.Unlock()
+	_, ok := l.l7EventsHttp[connkey]
+	if ok {
+		if exp.l7VisibilityController.GetTCNamefromPodNNSet(destinationPodNN) != "" || exp.l7VisibilityController.GetTCNamefromPodNNSet(sourcePodNN) != "" {
```

Comment:
Isn't the _direction_ of the traffic (when compared to the value of the L7visibility annotation) also important here, when deciding whether to export the L7 data? I wonder if the check should look more like this:
```go
if (exp.l7VisibilityController.GetTCNamefromPodNNSet(destinationPodNN) != "" && (direction == "ingress" || direction == "both") ) || (exp.l7VisibilityController.GetTCNamefromPodNNSet(sourcePodNN) != "" && (direction == "egress" || direction == "both))
```

Please let me know if I'm missing something

Reply:
We already complete this check in the L7Visibility controller, so the flow received should only contain the annotated direction data.

Reply:
I am not sure I understand. In the l7visibility controller, you call `addPodNNToSet` regardless of the direction value.

So let's say I have the following situation:
* a flow PodA -> PodB
* PodA has a visibility annotation with direction `ingress`
* PodB has no visibility annotation; however it has a L7 NP applied to it, so all its ingress traffic goes through Suricata

In this case, is it possible that we will send L7 flow information for the flow (PodA -> PodB), even though it was not requested by the user? Given that unless I am mistaken, `exp.l7VisibilityController.GetTCNamefromPodNNSet(sourcePodNN)` (where `sourcePodNN` is PodA) will return true?

In other words, this:
> We already complete this check in the L7Visibility controller

Only applies to TC flows created because of the L7Visibility annotation, not to those created because of L7NP?

Reply:
yes that could create unnecessary flows to reach flow aggregator, I have modified the function to match podNN and direction to the corresponding field for the flow to be added to the exporter.